File: src/main/java/com/coralblocks/coralbench/bench/SleepBenchmark.java
package com.coralblocks.coralbench.bench;

import com.coralblocks.coralbench.Bench;

public class SleepBenchmark {
	
	private final static void sleepFor(long nanos) {
		long time = System.nanoTime();
		while((System.nanoTime() - time) < nanos);
	}
	
	private final static void doSleep(Bench bench) {
		bench.mark();
		sleepFor(1000);
		bench.measure();
	}
	
	public static void main(String[] args) {
		
		int warmupIterations = 1000000;
		int measurementIterations = 1000000;
		
		Bench bench = new Bench(warmupIterations);
		
		while(bench.getCount() < measurementIterations + warmupIterations) {
			doSleep(bench);
		}
		
		bench.printResults();
		
		bench.reset(false); // false because we don't want to repeat warmup
		
		while(bench.getCount() < measurementIterations) {
			doSleep(bench);
		}
		
		bench.printResults();
	}
}

File: src/main/c/CoralBlocks/CoralBench/Bench/SleepBenchmark.cpp
#include "../Bench.cpp"

using namespace CoralBlocks::CoralBench;

struct timespec ts;

long get_nano_ts() {
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1000000000 + ts.tv_nsec;
}

void sleepFor(long nanos) {
    long total = get_nano_ts();
    while(get_nano_ts() - total < nanos);
}

void doSleep(Bench& bench) {
    bench.mark();
    sleepFor(1000);
    bench.measure();
}

int main() {

    const int warmupIterations = 1000000;
    const int measurementIterations = 1000000;

    Bench& bench = *(new Bench(warmupIterations)); // don't want to use -> to dereference

    while (bench.getCount() < measurementIterations + warmupIterations) {
        doSleep(bench);
    }

    bench.printResults();

    delete &bench;

    return 0;
}
